File: dotnet/EliteDangerousSdk/Planner/Colonization.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace EliteDangerousSdk.Planner
{
    public enum ColonyState
    {
        None = 0,
        Claimed = 1,
        BeaconPlaced = 2,
        PrimaryPortBuilding = 3,
        Active = 4,
        Failed = 5,
    }

    public static class ColonyStateNames
    {
        public static string GetName(ColonyState state) => state switch
        {
            ColonyState.None => "None",
            ColonyState.Claimed => "Claimed",
            ColonyState.BeaconPlaced => "Beacon Placed",
            ColonyState.PrimaryPortBuilding => "Primary Port Building",
            ColonyState.Active => "Active",
            ColonyState.Failed => "Failed",
            _ => "",
        };
    }

    public class ConstructionResource
    {
        public string Name { get; set; } = "";
        public string NameLocalised { get; set; } = "";
        public int RequiredAmount { get; set; }
        public int ProvidedAmount { get; set; }
        public int Payment { get; set; }
    }

    public class ConstructionSite
    {
        public string Id { get; set; } = "";
        public long MarketId { get; set; }
        public bool PrimaryPort { get; set; }
        public double ConstructionProgress { get; set; }
        public bool ConstructionComplete { get; set; }
        public bool ConstructionFailed { get; set; }
        public List<ConstructionResource> ResourcesRequired { get; set; } = new();
    }

    public class ColonySystem
    {
        public string SystemName { get; set; } = "";
        public string Architect { get; set; } = "";
        public ColonyState State { get; set; } = ColonyState.None;
        public List<ConstructionSite> ConstructionSites { get; set; } = new();
        public int CompletedStructures { get; set; }
        public int TotalSlots { get; set; }
    }

    public static class Colonization
    {
        private static int _nextId = 1;
        private static string GenId() => $"col-{_nextId++}";

        public static ConstructionSite CreateConstructionSite(long marketId, bool primaryPort = false)
        {
            return new ConstructionSite
            {
                Id = GenId(),
                MarketId = marketId,
                PrimaryPort = primaryPort,
            };
        }

        public static int GetResourceShortfall(ConstructionResource resource) =>
            Math.Max(0, resource.RequiredAmount - resource.ProvidedAmount);

        public static double GetTotalProgress(ConstructionSite site)
        {
            if (site.ResourcesRequired.Count == 0) return site.ConstructionProgress;
            var totalRequired = site.ResourcesRequired.Sum(r => r.RequiredAmount);
            if (totalRequired == 0) return site.ConstructionProgress;
            var totalProvided = site.ResourcesRequired.Sum(r => r.ProvidedAmount);
            return (double)totalProvided / totalRequired;
        }

        public static ConstructionSite ParseColonisationConstructionDepot(Dictionary<string, object> event_)
        {
            var resources = new List<ConstructionResource>();
            if (event_.TryGetValue("ResourcesRequired", out var raw) && raw is List<object> rawList)
            {
                foreach (var item in rawList)
                {
                    if (item is Dictionary<string, object> r)
                    {
                        var name = r.GetValueOrDefault("Name", "")?.ToString() ?? "";
                        resources.Add(new ConstructionResource
                        {
                            Name = name,
                            NameLocalised = r.GetValueOrDefault("Name_Localised", name)?.ToString() ?? name,
                            RequiredAmount = Convert.ToInt32(r.GetValueOrDefault("RequiredAmount", 0)),
                            ProvidedAmount = Convert.ToInt32(r.GetValueOrDefault("ProvidedAmount", 0)),
                            Payment = Convert.ToInt32(r.GetValueOrDefault("Payment", 0)),
                        });
                    }
                }
            }
            return new ConstructionSite
            {
                Id = GenId(),
                MarketId = Convert.ToInt64(event_.GetValueOrDefault("MarketID", 0L)),
                ConstructionProgress = Convert.ToDouble(event_.GetValueOrDefault("ConstructionProgress", 0.0)),
                ConstructionComplete = Convert.ToBoolean(event_.GetValueOrDefault("ConstructionComplete", false)),
                ConstructionFailed = Convert.ToBoolean(event_.GetValueOrDefault("ConstructionFailed", false)),
                ResourcesRequired = resources,
            };
        }
    }
}

File: python/elite_dangerous_sdk/planner/colonization.py
from dataclasses import dataclass, field
from enum import IntEnum
from typing import List


class ColonyState(IntEnum):
    NONE = 0
    CLAIMED = 1
    BEACON_PLACED = 2
    PRIMARY_PORT_BUILDING = 3
    ACTIVE = 4
    FAILED = 5


COLONY_STATE_NAMES = {
    ColonyState.NONE: "None",
    ColonyState.CLAIMED: "Claimed",
    ColonyState.BEACON_PLACED: "Beacon Placed",
    ColonyState.PRIMARY_PORT_BUILDING: "Primary Port Building",
    ColonyState.ACTIVE: "Active",
    ColonyState.FAILED: "Failed",
}


@dataclass
class ConstructionResource:
    name: str = ""
    name_localised: str = ""
    required_amount: int = 0
    provided_amount: int = 0
    payment: int = 0


@dataclass
class ConstructionSite:
    id: str = ""
    market_id: int = 0
    primary_port: bool = False
    construction_progress: float = 0.0
    construction_complete: bool = False
    construction_failed: bool = False
    resources_required: List[ConstructionResource] = field(default_factory=list)


@dataclass
class ColonySystem:
    system_name: str = ""
    architect: str = ""
    state: ColonyState = ColonyState.NONE
    construction_sites: List[ConstructionSite] = field(default_factory=list)
    completed_structures: int = 0
    total_slots: int = 0


_next_id = 1


def _gen_id() -> str:
    global _next_id
    result = f"col-{_next_id}"
    _next_id += 1
    return result


def create_construction_site(market_id: int, primary_port: bool = False) -> ConstructionSite:
    return ConstructionSite(
        id=_gen_id(),
        market_id=market_id,
        primary_port=primary_port,
    )


def get_resource_shortfall(resource: ConstructionResource) -> int:
    return max(0, resource.required_amount - resource.provided_amount)


def get_total_progress(site: ConstructionSite) -> float:
    if not site.resources_required:
        return site.construction_progress
    total_required = sum(r.required_amount for r in site.resources_required)
    if total_required == 0:
        return site.construction_progress
    total_provided = sum(r.provided_amount for r in site.resources_required)
    return total_provided / total_required


def parse_colonisation_construction_depot(event: dict) -> ConstructionSite:
    resources = []
    for r in event.get("ResourcesRequired", []):
        name = r.get("Name", "") or ""
        name_localised = r.get("Name_Localised", name) or name
        resources.append(ConstructionResource(
            name=name,
            name_localised=name_localised,
            required_amount=r.get("RequiredAmount", 0),
            provided_amount=r.get("ProvidedAmount", 0),
            payment=r.get("Payment", 0),
        ))
    return ConstructionSite(
        id=_gen_id(),
        market_id=event.get("MarketID", 0),
        construction_progress=float(event.get("ConstructionProgress", 0.0)),
        construction_complete=bool(event.get("ConstructionComplete", False)),
        construction_failed=bool(event.get("ConstructionFailed", False)),
        resources_required=resources,
    )
